docs/tutorials/developer-liftoff-rust/level-1/1.1-rust-lvl1.mdx

```diff
@@ -0,0 +1,227 @@
+
+```rust
+use candid::Principal;
+
+#[ic_cdk::query]
+fn whoami() -> Principal {
+    ic_cdk::caller()
+}
+
+// Export the Candid interface
+ic_cdk::export_candid!();
+```
+
+`query` and `update` methods can be hidden from the Candid exporter by setting `hidden` to equal `true`. If a method is hidden, it still exists in the canister but will not have a Candid interface generated through `export_candid!`
+
+```rust
+#[query(hidden = true)]
+fn query_function() {
+    // ...
+}
+```
+
+## Rust Wasm limitations
+
+WebAssembly (Wasm) is a portable binary format that runs in a virtual machine. ICP compiles canister code to standard Wasm using tools included in the `ic_cdk` that compile Rust code to the `wasm32-unknown-unknown` Wasm target.
+
+Rust relies on the Wasm binary and the associated Candid interface file to define and expose the canister's methods. In a `dfx.json` file that defines a Rust canister, you will need to specify additional fields to indicate where in the project these files are located:
+
+- `build` field: Specifies the command to produce the Wasm binary (using Cargo). If you've defined a package name in the project's `Cargo.toml` file, you can specify the `package` name instead.
+
+- `wasm` field: The path to the resulting Wasm file if `build` is defined.
+
+- `candid` field: The path to the Candid file that describes the canister's interface.
+
+For example:
+
+```json
+    "hello_world_rust": {
+    "type": "custom",
+    "build": "cargo build --target wasm32-unknown-unknown --package hello_world_rust --release",
+    "wasm": "target/wasm32-unknown-unknown/release/hello_world_rust.wasm",
+    "candid": "src/hello_world_rust/src/hello_world_rust.did"
+    },
+```
+
+Alternatively, if you have specified the `package` name in your Cargo.toml file, such as:
+
+```toml
+[package]
+name = "hello_world_rust"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+candid = "0.10"
+ic-cdk = "0.17"
+ic-cdk-timers = "0.11" # Feel free to remove this dependency if you don't need timers
+```
+
+Then you can omit the `build` and `wasm` fields in favor of the `package` and `type` fields:
+
+```json
+    "hello_world_rust": {
+      "candid": "src/hello_world_rust/hello_world_rust.did",
+      "package": "hello_world_rust",
+      "type": "rust"
+    },
+```
+
+## Rust Wasm limitations
+
+Canisters written in Rust have a set of limitations imposed by the Wasm compiler due to the fact that the IC does not provide a file system, network access, or other functionalities at the moment. These limitations include:
+
+- You cannot create threads. Instead, use [`ic_cdk::spawn`](https://docs.rs/ic_cdk/latest/ic_cdk/fn.spawn.html).
```

Comment:
The path to this is about to change - `latest` in this link should be replaced with `0.16.0` so it doesn't break when that happens, like below. But also, perhaps the 0.16 links should be updated to 0.17.

Reply:
0.18.3 has been released, so instead they should be 0.18.3 and use the updated link - https://docs.rs/ic-cdk/0.18.3/ic_cdk/futures/fn.spawn.html